File: puc_rio_cp/downloader/platforms/codeforces.py
import re

from .platform import Platform
from ..problem_info import ProblemInfo


class CodeforcesHandler(Platform):
    URL_PATTERNS = [
        r'.*codeforces\.com/contest/(\d+)/problem/([A-Z]\d*)',
        r'.*codeforces\.com/problemset/problem/(\d+)/([A-Z]\d*)',
        r'.*codeforces\.com/gym/(\d+)/problem/([A-Z]\d*)'
    ]
    
    def get_info_from_url(self, url: str) -> ProblemInfo:
        for pattern in self.URL_PATTERNS:
            if match := re.match(pattern, url):
                contest_id, letter = match.groups()
                folder_path = self._get_problem_directory('codeforces', contest_id)
                return ProblemInfo(
                    platform_name='codeforces',
                    problem_id=f"{contest_id}{letter}",
                    folder_path=folder_path,
                    file_name=letter.lower()
                )
        raise ValueError(f"Invalid Codeforces URL: {url}")

File: internal/platform/codeforces.go
package platform

import (
    "fmt"
    "net/url"
    "path"
    "regexp"
    "strings"
)

var (
    cfContestRe   = regexp.MustCompile(`(?i)codeforces\.com/(?:contest|gym)/(\d+)/problem/([A-Z]\d*)`)
    cfProblemsetRe = regexp.MustCompile(`(?i)codeforces\.com/problemset/problem/(\d+)/([A-Z]\d*)`)
)

type CodeforcesResolver struct{}

func (c CodeforcesResolver) Matches(rawURL string) bool {
    u, _ := url.Parse(rawURL)
    if u == nil { return false }
    host := strings.ToLower(u.Hostname())
    return strings.Contains(host, "codeforces.com") || strings.Contains(host, "codeforces")
}

func (c CodeforcesResolver) Resolve(basePath, rawURL string) (ProblemInfo, error) {
    if m := cfContestRe.FindStringSubmatch(rawURL); len(m) == 3 {
        contestID, letter := m[1], strings.ToLower(m[2])
        folder := path.Join(basePath, "codeforces", contestID)
        return ProblemInfo{Platform: "codeforces", ProblemID: contestID+letter, FolderPath: folder, FileName: letter}, nil
    }
    if m := cfProblemsetRe.FindStringSubmatch(rawURL); len(m) == 3 {
        contestID, letter := m[1], strings.ToLower(m[2])
        folder := path.Join(basePath, "codeforces", contestID)
        return ProblemInfo{Platform: "codeforces", ProblemID: contestID+letter, FolderPath: folder, FileName: letter}, nil
    }
    return ProblemInfo{}, fmt.Errorf("unrecognized codeforces url: %s", rawURL)
}

